clock/README.md

```diff
@@ -5,6 +5,362 @@ A Go time testing library for writing deterministic unit tests
+
+```go
+fired := false
+
+tmr := mClock.Afterfunc(time.Second, func() {
+  fired = true
+})
+mClock.Advance(time.Second)
+```
+
+When you call `Advance()` it immediately moves the clock forward the given amount, and triggers any
+tickers or timers that are scheduled to happen at that time. Any triggered events happen on separate
+goroutines, so _do not_ immediately assert the results:
+
+```go
+fired := false
+
+tmr := mClock.Afterfunc(time.Second, func() {
+  fired = true
+})
+mClock.Advance(time.Second)
+
+// RACE CONDITION, DO NOT DO THIS!
+if !fired {
+  t.Fatal("didn't fire")
+}
+```
+
+`Advance()` (and `Set()` for that matter) return an `AdvanceWaiter` object you can use to wait for
+all triggered events to complete.
+
+```go
+fired := false
+// set a test timeout so we don't wait the default `go test` timeout for a failure
+ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+
+tmr := mClock.Afterfunc(time.Second, func() {
+  fired = true
+})
+
+w := mClock.Advance(time.Second)
+err := w.Wait(ctx)
+if err != nil {
+  t.Fatal("AfterFunc f never completed")
+}
+if !fired {
+  t.Fatal("didn't fire")
+}
+```
+
+The construction of waiting for the triggered events and failing the test if they don't complete is
+very common, so there is a shorthand:
+
+```go
+w := mClock.Advance(time.Second)
+err := w.Wait(ctx)
+if err != nil {
+  t.Fatal("AfterFunc f never completed")
+}
+```
+
+is equivalent to:
+
+```go
+w := mClock.Advance(time.Second)
+w.MustWait(ctx)
+```
+
+or even more briefly:
+
+```go
+mClock.Advance(time.Second).MustWait(ctx)
+```
+
+### Advance only to the next event
+
+One important restriction on advancing the clock is that you may only advance forward to the next
+timer or ticker event and no further. The following will result in a test failure:
+
+```go
+func TestAdvanceTooFar(t *testing.T) {
+    ctx, cancel := context.WithTimeout(10*time.Second)
+    defer cancel()
+    mClock := quartz.NewMock(t)
+    var firedAt time.Time
+    mClock.AfterFunc(time.Second, func() {
+        firedAt := mClock.Now()
+    })
+    mClock.Advance(2*time.Second).MustWait(ctx)
+}
+```
+
+This is a deliberate design decision to allow `Advance()` to immediately and synchronously move the
+clock forward (even without calling `Wait()` on returned waiter). This helps meet Quartz's design
+goals of writing deterministic and easy to understand unit tests. It also allows the clock to be
+advanced, deterministically _during_ the execution of a tick or timer function, as explained in the
+next sections on Traps.
+
+Advancing multiple events can be accomplished via looping. E.g. if you have a 1-second ticker
+
+```go
+for i:=0; i<10; i++ {
+    mClock.Advance(time.Second).MustWait(ctx)
+}
+```
+
+will advance 10 ticks.
+
+If you don't know or don't want to compute the time to the next event, you can use `AdvanceNext()`.
+
+```go
+d, w := mClock.AdvanceNext()
+w.MustWait(ctx)
+// d contains the duration we advanced
+```
```

Comment:
It might be good to document when and why you would use `Advance` instead of `AdvanceNext`. `Advance` has basically the same limitation, except it can increment to times before any tick/timers.

To make `Advance` more useful, I think it would be good to have a `mClock.NextTick()`, this can allow you to write code like:

```go
for i := 0; i < 10; i++ {
	d := time.Second
	if nt := mClock.NextTick(); nt < d {
		d = nt
	}
	mClock.Advance(d).MustWait(ctx)
}
```

Thoughts? A use-case that is very hard to cover under `mClock.Advance` is:

```go
t := mClock.NewTicker(10 * time.Second)
for range t.C {
	// do job that takes nondeterministic amount of time.
	t.Reset(10 * time.Second)
}
```

PS. The argument could probably be made that this library is not meant for this kind of use-case, but then I question the utility of the `Advance` function.

Reply:
I do think the `NextTick()` call is useful, although I would probably call it `Next()` or perhaps `NextEvent()`, since it's not just tickers that are affected.

The use case I had in mind was being able to blow through timers and ticks to advance a certain duration.

```
desired := time.Minute
for desired > 0 {
    n := mClock.Next()
    if n > desired {
        mClock.Advance(desired).MustWait(ctx)
        break
    }
    mClock.Advance(n).MustWait(ctx)
    desired -= n
}
```

I'm not sure I understand the point of your first example.  Is it that you don't know the size of the ticks and you want to advance 10 seconds or 10 ticks, whichever comes first?

In terms of the second example, I'm pretty confident it works fine with `Advance()`: you can trap the `t.Reset()` call and control the amount of mocked time the job takes (the amount of real time should be irrelevant).

Reply:
Oh, and, I think `Advance()` is _usually_ preferable to `AdvanceNext()`, since it allows you to keep a mental tally of time passing during the test case.  But, not always.  There was even one test I refactored that I thought was clearer using `Set()`.

Reply:
I added `Next()` up stack from this, btw.